api/firmware/query.go

```diff
@@ -70,27 +75,35 @@ func (device *Device) rawQueryV7(msg []byte) ([]byte, error) {
 		break
 	}
 
-	var lastQueryTime time.Time
-
 	for {
 		switch status {
 		case hwwRspAck:
 			return payload, nil
 		case hwwRspBusy:
 			return nil, errp.New("unexpected hwwRspBusy response")
 		case hwwRspNack:
-			if lastQueryTime.IsZero() {
+			lastQueryTimes = lastQueryTimes.Prev()
+			if lastQueryTimes.Prev().Value == nil {
 				device.log.Debug("unexpected NACK response in first loop iteration")
 			} else {
-				device.log.Debug(
-					fmt.Sprintf(
-						"unexpected NACK response; last successful query was %v ago",
-						time.Since(lastQueryTime)))
+				logStr := "unexpected NACK response; last successful retry query intervals (newest on top) were:\n"
```

Comment:
I'd turn the newlines into semicolons. Log lines are usually one liners. so one can for example grep the date/time at the beginning of a line or filter groups or other fields and not miss any output.